File: june14/june27_HM2/generichm.cpp
#include<iostream>
#include<list>
#include<string>
using namespace std;

template<class k,class v>
class hm
{
    private:
        class hmnode
        {
            public:
                k key;
                v value;
                hmnode(k key,v value)
                {
                    this->key=key;
                    this->value=value;
                }
        };
        int n;
        int N;
        list<hmnode>* buckets=NULL;

        int hashFn(k key)
        {
            int hc=hash<k>()(key);
            hc=abs(hc);
            return hc%N;
        }

        typename list<hmnode>::iterator getDataIndex(int bi,k key)
        {
            for(auto itr=buckets[bi].begin();itr!=buckets[bi].end();itr++)
            {
                if(itr->key==key)
                    return itr;
            }
            return buckets[bi].end();
        }
        void rehash()
        {
            int oldN=N;
            list<hmnode>* oba=buckets;
            N=N*2;
            buckets=new list<hmnode>[N];
            n=0;

            for(int bi=0;bi<oldN;bi++)
            {
                for(auto ditr=oba[bi].begin();ditr!=oba[bi].end();ditr++)
                    (*this)[ditr->key] = ditr->value;
            }
        }
    public:
        hm()
        {
            n=0;
            N=4;
            buckets=new list<hmnode>[N];
        }

        int count(k key)
        {
            int bi=hashFn(key);
            auto ditr=getDataIndex(bi,key);
            if(ditr==buckets[bi].end())
                return 0;
            else
                return 1;
        }

        v& operator[](const k& key)
        {
            int bi=hashFn(key);
            auto ditr=getDataIndex(bi,key);
            if(ditr==buckets[bi].end())
            {
                hmnode node(key,v());
                buckets[bi].push_back(node);
                n++;

                double lambda=n*1.0/N;
                if(lambda>2.0)
                {
                    rehash();
                    bi=hashFn(key);
                    ditr=getDataIndex(bi,key);
                }
                else
                {
                    return buckets[bi].back().value;
                }    
            }
            return ditr->value;
        }
        void erase(k key)
        {
            int bi=hashFn(key);
            auto ditr=getDataIndex(bi,key);
            if(ditr!=buckets[bi].end())
            {
                buckets[bi].erase(ditr);
                n--;
            }
        }
        int size()
        {
            return n;
        }
        void display()
        {
            cout << "-------------------------" << endl;
            for(int bi=0;bi<N;bi++)
            {
                cout<< bi << " -> " ;
                for(auto ditr=buckets[bi].begin();ditr!=buckets[bi].end();ditr++)
                {
                    cout<< "[" << ditr->key << "=" << ditr->value << "]";
                }
                cout<< endl;
            }
            cout << "-------------------------" << endl;
        }
};  

class point
{
    public:
        int x;
        int y;
        point()
        {
        }
        point(int x,int y)
        {
            this->x=x;
            this->y=y;
        }

        bool operator==(const point& p)const
        {
            return this->x==p.x && this->y==p.y;
        }

        friend ostream& operator<<(ostream& out,const point& p)
        {
            out<< p.x << " " << p.y;
            return out;
        }
};
namespace std
{
    template<>
    struct hash<point>
    {
        public:
            int operator()(const point& p)
            {
                return p.x+p.y;
            }
    };
}
int main(int argc,char** argv)
{
    // hm<string,int> map;
    // map["India"]=100;
    // map["Pak"]=80;
    // cout<< map.count("India") << " " << map.count("Pak") <<endl;
    // cout<< map["India"] << " " << map["Pak"] << endl;
    // cout<< map.count("Sus") << endl;
    // cout<< map["Sus"] << endl;
    // cout<< map.count("Sus") << endl;

    // map.erase("Sus");
    // cout<< map.count("Sus") << endl;
    // map["India"]=135;
    // map["Norway"]=10;
    // map["China"]=200;
    // map["US"]=123;
    // map["UK"]=150;
    // map["Bangladesh"]=12;
    // map["California"]=120;
    // map["Afghanistan"]=1;
    // map.display();

    hm<point,string> map;
    point p1(0, 2);
    point p2(3, 4);
    point p3(2, 2);
    point p4(4, 7);
    point p5(2, 1);
    point p6(2, 1);

    map[p1]= "Delhi";
    map[p2]="Chandigarh";
    map[p3]="Mumbai";
    map[p4]="Shimla";
    map[p5]="Murthal";

    cout << p6 << endl;

    // cout << "abc" << 123 << "abc" << endl;
    // cout.operator<<("abc").operator<<(123).operator<<("abc");
    // above two lines have same functionality
    map.display();

}

File: june14/june23_HM/generichm.java
import java.util.*;
public class generichm<k,v>{
    class HMNode{
        k key;
        v value;
        HMNode(k key,v value){
            this.key=key;
            this.value=value;
        }
    }
    LinkedList<HMNode>[] buckets;
    int n;//total no. of elements
    int N;//no. of buckets
    generichm(){
        N=4;
        buckets=new LinkedList[N];
        for(int i=0;i<buckets.length;i++)
            buckets[i]=new LinkedList<>();
        n=0;
    }
    private int hashfn(k key){
        int hc=key.hashCode();
        hc=Math.abs(hc);
        return hc%N;
    }
    private int getDataIndex(int bi,k key){
        int counter=0;
        for(HMNode node:buckets[bi]){
            if(node.key.equals(key))
                return counter;
            counter++;
        }
        return -1;
    }
    private void rehash(){
        LinkedList<HMNode>[] oba=buckets;
        N=N*2;
        buckets=new LinkedList[N];
        for(int i=0;i<buckets.length;i++)
            buckets[i]=new LinkedList<>();
        n=0;

        for(int bi=0;bi<oba.length;bi++){
            for(HMNode node: oba[bi])
                put(node.key, node.value);
        }
    }
    public void put(k key,v value){
        int bi=hashfn(key);
        int di=getDataIndex(bi, key);
        if(di==-1){
            HMNode n2a=new HMNode(key, value);
            buckets[bi].add(n2a);
            n++;
        }else{
            HMNode n2u=buckets[bi].get(di);
            n2u.value=value;
        }
        double lambda=n*1.0/N;
        if(lambda>2.0)
            rehash();
    }
    public boolean containsKey(k key){
        int bi=hashfn(key);
        int di=getDataIndex(bi, key);
        if(di==-1)
            return false;
        else
            return true;
    }
    public v get(k key){
        int bi=hashfn(key);
        int di=getDataIndex(bi, key);
        if(di==-1)
            return null;
        else{
            HMNode n2g=buckets[bi].get(di);
            return n2g.value;
        }
    }
    public v remove(k key){
        int bi=hashfn(key);
        int di=getDataIndex(bi, key);
        if(di==-1)
            return null;
        else{
            HMNode n2r=buckets[bi].remove(di);
            n--;
            return n2r.value;
        }
    }
    public int size(){
        return n;
    }
    public ArrayList<k> keyset(){
        ArrayList<k> keys=new ArrayList<>();
        for(int i=0;i<buckets.length;i++){
            for(HMNode node:buckets[i])
                keys.add(node.key);
        }
        return keys;
    }
    public void display(){
        System.out.println("-------------------------");
        for(int i=0;i<buckets.length;i++){
            System.out.print(i+" -> ");
            for(HMNode node:buckets[i])
                System.out.print("["+node.key+"="+node.value+"]");
            System.out.println();
        }
        System.out.println("-------------------------");
    }

    static class Point{
        int x;
        int y;
        Point(int x,int y){
            this.x=x;
            this.y=y;
        }
        public int hashCode(){
            return x+y;
        }
        public boolean equals(Object o){
            Point po=(Point)o;
            return this.x==po.x && this.y==po.y;
        }
        public String toString(){
            return this.x+" "+this.y;
        }
    }
    public static void main(String[] args) {
        // generichm<String,Integer> hm=new generichm<>();
        // hm.put("India", 200);
        // hm.put("Pak", 30);
        // hm.put("China", 250);
        // System.out.println(hm.get("China"));
        // System.out.println(hm.containsKey("Pak"));
        // hm.remove("Pak");
        // System.out.println(hm.containsKey("Pak"));
        // hm.put("India", 135);
        // hm.put("US", 260);
        // hm.put("UK", 120);
        // hm.put("Bangladesh", 20);
        // hm.put("California", 90);
        // hm.put("Pak", 10);
        // hm.put("Afghanistan", 15);
        // hm.put("Norway", 23);
        // hm.display();

        generichm<Point,String> map=new generichm<>();
        Point p1=new Point(0, 2);
        Point p2=new Point(3, 4);
        Point p3=new Point(2, 2);
        Point p4=new Point(4, 7);
        Point p5=new Point(2, 1);
        Point p6=new Point(2, 1);

        map.put(p1, "Delhi");
        map.put(p2, "Chandigarh");
        map.put(p3, "Mumbai");
        map.put(p4, "Shimla");
        map.put(p5, "Murthal");

        System.out.println(map.get(p6));
        System.out.println(p5);
        map.display();
    }
}